Begin experimental support for pin reborrowing

eholk opened the pull request:
This commit adds basic support for reborrowing `Pin` types in argument position. At the moment it only supports reborrowing `Pin<&mut T>` as `Pin<&mut T>` by inserting a call to `Pin::as_mut()`, and only in argument position (not as the receiver in a method call).

This PR makes the following example compile:

```rust
#![feature(pin_ergonomics)]

fn foo(_: Pin<&mut Foo>) {
}

fn bar(mut x: Pin<&mut Foo>) {
    foo(x);
    foo(x);
}
```

Previously, you would have had to write `bar` as:

```rust
fn bar(mut x: Pin<&mut Foo>) {
    foo(x.as_mut());
    foo(x);
}
```

Tracking:

- #130494 

r? @compiler-errors

eholk committed:
Begin experimental support for pin reborrowing

This commit adds basic support for reborrowing `Pin` types in argument
position. At the moment it only supports reborrowing `Pin<&mut T>` as
`Pin<&mut T>` by inserting a call to `Pin::as_mut()`, and only in
argument position (not as the receiver in a method call).

compiler-errors commented on compiler/rustc_mir_build/src/thir/cx/expr.rs:
```diff
@@ -146,6 +147,50 @@ impl<'tcx> Cx<'tcx> {
                 ExprKind::RawBorrow { mutability, arg: self.thir.exprs.push(expr) }
             }
             Adjust::DynStar => ExprKind::Cast { source: self.thir.exprs.push(expr) },
+            Adjust::ReborrowPin(AutoBorrow::Ref(region, m)) => {
```
I think lowering to `as_mut` might be problematic. Specifically, because of the `Deref` bound and the "generality" of `as_mut`, this code breaks:

```
#![feature(pin_ergonomics)]

use std::pin::Pin;

fn shorter<'b, T: 'b>(x: Pin<&'b mut T>) {}

fn test<'a: 'b, 'b, T: 'a>(mut x: Pin<&'a mut T>) {
    shorter::<'b>(x);
}
```

Because `Pin::as_mut` takes a ref to the *pin*. I think you want to do some sort of desugaring that is more analogous to:

```
fn shorten<'a: 'b, 'b, T>(_: Pin<&'a mut T>) -> Pin<&'b mut T>
```

eholk replied:
Okay, I just pushed some changes that desugar the reborrow as `Pin { __pointer: &mut *x.__pointer }` which solves this issue.

compiler-errors reviewed (changes requested):
Left some thoughts. Specifically, aside from the nits, there's a problem with the desugaring that makes it not as general as when the feature gate is *disabled* due to the specific lowering to a call of `Pin::as_mut`. I think that should be changed.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/10928997481/job/30338690828) [(plain)](https://github.com/rust-lang/rust/commit/7b7992fbcf5b8d9af945e32eac241252baa46fab/checks/30338690828/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 2.820 Building wheels for collected packages: reuse
#16 2.821   Building wheel for reuse (pyproject.toml): started
#16 3.071   Building wheel for reuse (pyproject.toml): finished with status 'done'
#16 3.072   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132715 sha256=dfa09868353292d98f811d3efdb0d54d07389e808efc71d68e3b93c514bf8bec
#16 3.072   Stored in directory: /tmp/pip-ephem-wheel-cache-5f58frc4/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#16 3.075 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#16 3.461 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#16 3.461 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#16 DONE 3.5s
---
 Documenting rustc_lexer v0.0.0 (/checkout/compiler/rustc_lexer)
error: unclosed HTML tag `Ptr`
   --> compiler/rustc_middle/src/ty/adjustment.rs:108:19
    |
108 |     /// Take a Pin<Ptr> and call either as_mut() or as_ref() to get a Pin<&mut T> or Pin<&T>.
    |
    = note: `-D rustdoc::invalid-html-tags` implied by `-D warnings`
    = help: to override `-D warnings` add `#[allow(rustdoc::invalid_html_tags)]`
help: try marking as source code
help: try marking as source code
    |
108 |     /// Take a `Pin<Ptr>` and call either as_mut() or as_ref() to get a Pin<&mut T> or Pin<&T>.

 Documenting rustc_serialize v0.0.0 (/checkout/compiler/rustc_serialize)
 Documenting rustc_arena v0.0.0 (/checkout/compiler/rustc_arena)
error: could not document `rustc_middle`
error: could not document `rustc_middle`
warning: build failed, waiting for other jobs to finish...
Command has failed. Rerun with -v to see more details.
  local time: Wed Sep 18 20:06:33 UTC 2024
  network time: Wed, 18 Sep 2024 20:06:33 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

compiler-errors commented:
Thanks for making this pretty bite-sized though. I was somewhat afraid you were going to put up a huge pin ergonomics PR; I'm happy to review this work, but if possible, I'd like to have more PRs that are smaller than fewer PRs that are larger (I know I am [a huge hypocrite](https://github.com/rust-lang/rust/pull/130514) for doing that exact thing, so if it's not possible, then I understand lol).

compiler-errors commented on tests/ui/async-await/pin-reborrow-self.rs:
```diff
@@ -0,0 +1,24 @@
+
+fn bar(x: Pin<&mut Foo>) {
+    x.foo();
+    x.foo(); // for this to work we need to automatically reborrow,
```
Not necessarily saying it needs to be done, but it would be cool to at least understand why it's harder than I think it is haha

eholk replied:
This is actually the case I wanted to work, I just didn't manage to find the right place to make it happen. I was actually kind of surprised that by this point we hadn't desugared everything to UFCS so the code paths would be the same.

In the interest of keeping PRs small and atomic, I'd rather add method calls in the next PR at this point.

eholk commented:
> Thanks for making this pretty bite-sized though. I was somewhat afraid you were going to put up a huge pin ergonomics PR; I'm happy to review this work, but if possible, I'd like to have more PRs that are smaller than fewer PRs that are larger (I know I am [a huge hypocrite](https://github.com/rust-lang/rust/pull/130514) for doing that exact thing, so if it's not possible, then I understand lol).

Thanks for the comments! I'm a fan of going for the smallest atomic change I can do, but sometimes the smallest atomic change is quite large. The nice thing about the pin ergonomics experiment is that it's around 4-6 independent features and half of them should be pretty small.

eholk commented on tests/ui/async-await/pin-reborrow-arg.rs:
```diff
@@ -0,0 +1,27 @@
+fn foo(_: Pin<&mut Foo>) {
+}
+
+fn bar(mut x: Pin<&mut Foo>) {
```
I opened #130535 to track this.

eholk replied:
As to why this happens, it's because I desugar the reborrow as `Pin::as_mut(&mut x)`, and the inner `& mut x` needs `x` to be mutable.

eholk replied:
This is no longer needed after I changed it to borrow the internals directly.

eholk commented on compiler/rustc_hir_typeck/src/expr_use_visitor.rs:
```diff
@@ -780,6 +780,20 @@ impl<'tcx, Cx: TypeInformationCtxt<'tcx>, D: Delegate<'tcx>> ExprUseVisitor<'tcx
+                    };
+                    self.delegate.borrow_mut().borrow(&place_with_id, place_with_id.hir_id, bk);
+
+                    self.walk_autoref(expr, &place_with_id, autoref);
```
Sounds good to me.  I'll pull it out.

eholk committed:
Apply code review suggestions

eholk committed:
Allow shortening reborrows

Generating a call to `as_mut()` let to more restrictive borrows than
what reborrowing usually gives us. Instead, we change the desugaring to
reborrow the pin internals directly which makes things more expressive.

eholk committed:
Add a test case to make sure we don't reborrow twice

eholk committed:
pin_ergonomics: allow reborrowing as Pin<&T>

eholk commented:
It went ahead and added support for reborrowing `Pin<&mut T>` and `Pin<&T>` as `Pin<&T>`, along with adding a test to make sure you can't reborrow `Pin<&T>` as `Pin<&mut T>`.

compiler-errors commented:
@bors r+ rollup

bors commented:
:pushpin: Commit a18800f8078f3fb7b986183b53c34a6a1b921d63 has been approved by `compiler-errors`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors a18800f8078f3fb7b986183b53c34a6a1b921d63 -->
<!-- homu: {"type":"Approved","sha":"a18800f8078f3fb7b986183b53c34a6a1b921d63","approver":"compiler-errors","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
